pkg/network/netlink/consumer.go

```diff
@@ -0,0 +1,410 @@
+	sampler, _ := GenerateBPFSampler(c.samplingRate)
+	err = c.socket.SetBPF(sampler)
+	if err != nil {
+		atomic.StoreInt64(&c.samplingPct, 0)
+		return fmt.Errorf("failed to attach BPF filter: %w", err)
+	}
+
+	return nil
+}
+
+// receive netlink messages and flushes them to the Event channel.
+// This method gets called in two different contexts:
+//
+// - During system-probe startup, when we're loading all entries from the Conntrack table.
+// In this case c.streaming attribute is false, and once we detect the end of the multi-part
+// message we stop calling socket.Receive() and close the output channel to signal upstream
+// consumers we're done.
+//
+// - When we're streaming new connection events from the netlink socket. In this case, `c.streaming`
+// attribute is true, and only when we detect an EOF we close the output channel.
+// It's also worth noting that in the event of an ENOBUF error, we'll re-create a new netlink socket,
+// and attach a BPF sampler to it, to lower the the read throughput and save CPU.
+func (c *Consumer) receive(output chan Event) {
+ReadLoop:
+	for {
+		buffer := c.pool.Get().(*[]byte)
+		msgs, err := c.socket.ReceiveInto(*buffer)
+
+		if err != nil {
+			switch socketError(err) {
+			case errEOF:
+				// EOFs are usually indicative of normal program termination, so we simply exit
+				return
+			case errENOBUF:
+				atomic.AddInt64(&c.enobufs, 1)
+			default:
+				atomic.AddInt64(&c.readErrors, 1)
+			}
+		}
+
+		throttlingErr := c.throttle(len(msgs))
+		if throttlingErr != nil {
+			return
+		}
+
+		// Messages with error codes are simply skipped
+		for _, m := range msgs {
+			if err := checkMessage(m); err != nil {
+				atomic.AddInt64(&c.msgErrors, 1)
+				continue ReadLoop
+			}
+		}
+
+		// Skip multi-part "done" messages
+		multiPartDone := len(msgs) > 0 && msgs[len(msgs)-1].Header.Type == netlink.Done
+		if multiPartDone {
+			msgs = msgs[:len(msgs)-1]
+		}
+
+		output <- c.eventFor(msgs, buffer)
+
+		// If we're doing a conntrack dump we terminate after reading the multi-part message
+		if multiPartDone && !c.streaming {
+			return
+		}
+	}
+}
+
+func (c *Consumer) eventFor(msgs []netlink.Message, buffer *[]byte) Event {
+	return Event{
+		msgs:   msgs,
+		buffer: buffer,
+		pool:   c.pool,
+	}
+}
+
+// throttle ensures that the read throughput from the socket stays below
+// the configured maxMessagePerSecond
+func (c *Consumer) throttle(numMessages int) error {
+	// We don't throttle the socket during initialization
+	// (when we dump the whole Conntrack table)
+	if !c.streaming {
+		return nil
+	}
+
+	c.breaker.Tick(numMessages)
+	if !c.breaker.IsOpen() {
+		return nil
+	}
+	atomic.AddInt64(&c.throttles, 1)
+
+	// Close current socket
+	c.socket.Close()
+
+	// Create new socket with the desired sampling rate
+	// We calculate the required sampling rate to reach the target maxMessagesPersecond
+	samplingRate := float64(c.rateLimit) * overshootFactor * c.samplingRate / float64(c.breaker.Rate())
```

Comment:
I'm curious why you wrote it in this order, I've been thinking about the calculation in this order:


```
float64(c.rateLimit) / float64(c.breaker.Rate()) * c.samplingrate * overshootFactor
```


Since you're left to right:
a) getting what percent of messages you actually want
b) getting that percent of the existing samplingRate
c) adjust

Reply:
I updated the order to make it more readable.